### PR TITLE
[Merged by Bors] - feat: port Analysis.Calculus.BumpFunctionInner

### PR DESCRIPTION
**Pull request by urkud**
---
The auxiliary definitions I deleted are never used outside of this file.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/BumpFunctionInner.lean [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+
+end Real
+
+variable {E X : Type _}
+
+/-- `f : ContDiffBump c`, where `c` is a point in a normed vector space, is a
+bundled smooth function such that
+
+- `f` is equal to `1` in `Metric.closedBall c f.r`;
+- `support f = Metric.ball c f.R`;
```
Comment (semorrison) [inline Mathlib/Analysis/Calculus/BumpFunctionInner.lean:259]:
```suggestion
- `f` is equal to `1` in `Metric.closedBall c f.rIn`;
- `support f = Metric.ball c f.rOut`;
```

File: Mathlib/Analysis/Calculus/BumpFunctionInner.lean [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+#align exp_neg_inv_glue.nonneg expNegInvGlue.nonneg
+
+-- porting note: new lemma
+@[simp] theorem zero_iff_nonpos : expNegInvGlue x = 0 ↔ x ≤ 0 :=
+  ⟨fun h ↦ not_lt.mp fun h' ↦ (pos_of_pos h').ne' h, zero_of_nonpos⟩
+
+/-!
+### Smoothness of `expNegInvGlue`
+
+Porting note: Yury Kudryashov rewrote the proof whlie porting, generalizing auxiliary lemmas and
```
Comment (semorrison) [inline Mathlib/Analysis/Calculus/BumpFunctionInner.lean:92]:
```suggestion
Porting note: Yury Kudryashov rewrote the proof while porting, generalizing auxiliary lemmas and
```

File: Mathlib/Analysis/Calculus/BumpFunctionInner.lean [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+
+open Polynomial Real Filter Set Function
+
+open scoped Polynomial
+
+/-- `expNegInvGlue` is the real function given by `x ↦ exp (-1/x)` for `x > 0` and `0`
+for `x ≤ 0`. It is a basic building block to construct smooth partitions of unity. Its main property
+is that it vanishes for `x ≤ 0`, it is positive for `x > 0`, and the junction between the two
+behaviors is flat enough to retain smoothness. The fact that this function is `C^∞` is proved in
+`exp_neg_inv_glue.smooth`. -/
```
Comment (semorrison) [inline Mathlib/Analysis/Calculus/BumpFunctionInner.lean:59]:
```suggestion
`expNegInvGlue.smooth`. -/
```

File: Mathlib/Analysis/Calculus/BumpFunctionInner.lean [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+* `f : ContDiffBump c`, where `c` is a point in a real vector space, is
+  a bundled smooth function such that
+
+  - `f` is equal to `1` in `Metric.closedBall c f.r`;
+  - `support f = Metric.ball c f.R`;
+  - `0 ≤ f x ≤ 1` for all `x`.
+
+  The structure `ContDiffBump` contains the data required to construct the
+  function: real numbers `r`, `R`, and proofs of `0 < r < R`. The function itself is available
+  through `coe_fn`.
```
Comment (semorrison) [inline Mathlib/Analysis/Calculus/BumpFunctionInner.lean:40]:
```suggestion
  - `f` is equal to `1` in `Metric.closedBall c f.rIn`;
  - `support f = Metric.ball c f.rOut`;
  - `0 ≤ f x ≤ 1` for all `x`.

  The structure `ContDiffBump` contains the data required to construct the
  function: real numbers `rIn`, `rOut`, and proofs of `0 < rIn < rOut`. The function itself is available
  through `CoeFun`.
```

**Comment by semorrison** [comment]
bors d+

File: Mathlib/Analysis/Calculus/BumpFunctionInner.lean [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+* `f : ContDiffBump c`, where `c` is a point in a real vector space, is
+  a bundled smooth function such that
+
+  - `f` is equal to `1` in `Metric.closedBall c f.rIn`;
+  - `support f = Metric.ball c f.rOut`;
+  - `0 ≤ f x ≤ 1` for all `x`.
+
+  The structure `ContDiffBump` contains the data required to construct the
+  function: real numbers `rIn`, `rOut`, and proofs of `0 < rIn < rOut`. The function itself is available
+  through `CoeFun`.
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/BumpFunctionInner.lean:40]:
```suggestion
  function: real numbers `rIn`, `rOut`, and proofs of `0 < rIn < rOut`. The function itself is
  available through `CoeFun`.
```

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**